Add separate cache for getPkgConfigDb

andreabedini opened the pull request:
Querying pkg-config for the version of every module can be a very expensive operation on some systems. This change adds a separate, per-project, cache for pkgConfigDB; reducing the cost from "every plan change" to "every pkg-config-db change per project".

No input key is required since getPkgConfigDb already specifies the directories to monitor for changes. These are obtained from pkg-config itself as `pkg-config --variable pc_path pkg-config` as documented in pkg-config(1).

A notice is presented to the user when refreshing the packagedb.

Closes #8930 and subsumes #9360.

changelog entry to follow

geekosaur reviewed (commented):
Would it be worth adding a test that ran cabal twice and verified that the cached `pkg-config` output is used the second time?

andreabedini commented:
> Would it be worth adding a test that ran cabal twice and verified that the cached `pkg-config` output is used the second time?

I think it would be, I am not sure whether we have anything similar already. I'll check.

geekosaur commented:
Not for caching, but there's a test that verifies that a package reinstall causes a warning IIRC.

gbaz commented:
this looks excellent, thanks!

chreekat commented on changelog.d/pr-9422:
```diff
@@ -0,0 +1,12 @@
+
+description: {
+  Querying pkg-config for the version of every module can be a very expensive
+  operation on some systems. This change adds a separate, per-project, for
```
A separate *cache*?

(I think this word is also missing from the PR description.)

andreabedini replied:
Cut&pasted from the commit message which is also wrong :sob: :relaxed:

mpickering commented:
Can you describe the cache invalidation strategy? 

Does the cache get invalidated when `PKG_CONFIG_PATH` environment variable gets changed?

It would be good to add some tests testing the different invalidation mechanisms.

andreabedini commented:
> Can you describe the cache invalidation strategy?

Yes, I am relying on what `getPkgConfigDb` does currently:

```
getPkgConfigDb :: Verbosity -> ProgramDb -> Rebuild PkgConfigDb
getPkgConfigDb verbosity progdb = do
  dirs <- liftIO $ getPkgConfigDbDirs verbosity progdb
  -- Just monitor the dirs so we'll notice new .pc files.
  -- Alternatively we could monitor all the .pc files too.
  traverse_ monitorDirectoryStatus dirs
  liftIO $ readPkgConfigDb verbosity progdb
```

It obtains PKG_CONFIG_PATH and for monitors each directory for changes. As the comment says, it does not monitor each individual files.

Perhaps it could be worth monitoring the result of `getPkgConfigDbDirs` itself?

> It would be good to add some tests testing the different invalidation mechanisms.

I agree. Maybe `cabal-install/tests/UnitTests/Distribution/Client/Store.hs` can serve as a starting point.

andreabedini commented:
I did some rework and added tests but one of the tests (`ExtraProgPath`) seems to have picked up a small difference in behaviour.

andreabedini commented on cabal-testsuite/PackageTests/ExtraProgPath/setup.out:
This change looks suspicious

andreabedini replied:
I figured it out 😊

andreasabel commented on cabal-testsuite/PackageTests/Backpack/Includes2/cabal-external-target.out:
```diff
@@ -1,4 +1,5 @@
 # cabal v2-build
+Querying pkg-config database...
```
We could print it only when this step takes a long time (for some definition of long).

andreabedini replied:
This a good idea but I think it should be a separate PR :relaxed:

andreabedini replied:
What is the problem of touching every testcase? Let me understand, do you find unacceptable to print this message?
It appears only when pkg-config is actually called, which, because of the cache, does not happen at every invocation.

andreasabel replied:
> It appears only when pkg-config is actually called, which, because of the cache, does not happen at every invocation.

Yes, but apparently `pkg-config` is called even when no external package is involved, it seems.  Like if I look at a random test, like https://github.com/haskell/cabal/tree/f97fb81a12e0daadc884ca004419d8354de93253/cabal-testsuite/PackageTests/MultiRepl/CabalTooOld , then the sources of this test do not mention anything with `pkg-config`, but still the message appears.

I think `cabal` is already spamming the user with lots of irrelevant information that is printed all the time, and we should strive to reduce this and let cabal communicate the _relevant_ things.

andreabedini replied:
Alright, I bumped the message to `info`.

andreabedini commented:
@mpickering do you think the tests are sufficient? I am checking the search path (for the pkg-config executable) and PKG_CONFIG_PATH. I don't think it is a perfect solution but it seems to be the approach taken elsewhere.

michaelpj commented:
So from a Nix perspective, the thing that should force cache invalidation is that we have a different `pkg-config` executable in the two cases. I'm not sure how we should detect that - I guess we could include the fully-resolved path to the executable in the cache key?

andreabedini commented:
I did include the resolved pkg-config location in the key for a bit but eventually removed it because
1. the code flow made it a bit awkward. Both `readPkgConfigDb` and `getPkgConfigDbDirs` call `needProgram` but they don't pass back the resulting programdb so the result is lost.
2. It seems like most of the time we limit ourself to look at the programsearch path

The options I had were
1. call `needProgram` a third time
2. refactor more code in cabal-install-solver
3. rely on program search path

@fgaz do you mean pkgconfig-depends? zlib (the haskell package) does not use pkg-config by default.

andreabedini commented:
Give the analysis above, I think this PR is complete. Nevertheless I would like to be sure there is a workaround for the situation that @fgaz describes above.

andreabedini commented:
I realised I didn't need any refactor after all! @fgaz could you give it another go?

andreabedini commented:
@jasagredo does [3ae0488](https://github.com/haskell/cabal/pull/9422/commits/3ae04882fde1b344150581eebca7291800ec3dd0) look alright to you? I tried to follow what you did in #9527

andreabedini commented:
```
Actual output differs from expected:

stderr:
--- /home/runner/work/cabal/cabal/cabal-testsuite/PackageTests/ExtraProgPath/setup.dist/setup.out.normalized	2024-01-24 14:53:40.341551022 +0000
+++ /home/runner/work/cabal/cabal/cabal-testsuite/PackageTests/ExtraProgPath/setup.dist/setup.comp.out.normalized	2024-01-24 14:53:40.341551022 +0000
@@ -1,8 +1,6 @@
 # cabal v2-build
 Warning: cannot determine version of <ROOT>/./pkg-config :
 ""
-Warning: cannot determine version of <ROOT>/./pkg-config :
-""
 Resolving dependencies...
 Error: [Cabal-7107]
 Could not resolve dependencies:
*** Exception: ExitFailure 1

*** unexpected failure for PackageTests/ExtraProgPath/setup.test.hs
```

:monocle_face:

fgaz commented:
@michaelpj

> So from a Nix perspective, the thing that should force cache invalidation is that we have a different `pkg-config` executable in the two cases

the pkg-config executable is the same in all my shells. What changes is `$PKG_CONFIG_PATH_FOR_TARGET`, confirmed by https://github.com/NixOS/nixpkgs/blob/55ca7939944136731b34316a4c6fc2d4a11da122/pkgs/build-support/pkg-config-wrapper/pkg-config-wrapper.sh

---

@andreabedini I tried again and the result is identical

michaelpj commented:
> the pkg-config executable is the same in all my shells. What changes is the non-standard variable $PKG_CONFIG_PATH_FOR_TARGET

andreabedini commented:
I don't have much time to dedicate to this so I would appreciate if anybody could help pushing this over the line. It's complete but I had issues with the tests.

jasagredo commented:
So I tried to run this on Windows again. Some remarks:
1. There are symlinks here. I would advise against that, they do not work on windows, one has to do weird magic tricks to support them.
2. A test is actually failing: `PackageTests\CCompilerOverride\setup.test.hs`. I will look into it eventually. It says it cannot find the C compiler.
3. My `./validate.sh` process dies abruptly with this same error I reported time ago: https://github.com/haskell/cabal/issues/9571

jasagredo commented:
I got this error on my machine:
```
fatal: 'C:\msys64\tmp\repos-67360\src' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
Unit Tests
  UnitTests.Distribution.Client.Get
    forkPackages, network tests
      git clone:                                                                                                                       FAIL (2.06s)
        tests\UnitTests\Distribution\Client\Get.hs:219:
        expected a file to exist: C:\msys64\tmp\repos-67360\zlib1/zlib.cabal
        Use -p '/git clone/' to rerun this test only.

1 out of 543 tests failed (77.03s)
<<< C:\Users\Javier\code\cabal\dist-newstyle-validate-ghc-9.8.2\build\x86_64-windows\ghc-9.8.2\cabal-install-3.11.0.0\t\unit-tests\build\unit-tests\unit-tests.exe -j1 --hide-successes (87/104 sec, 1)
<<< C:\Users\Javier\code\cabal\dist-newstyle-validate-ghc-9.8.2\build\x86_64-windows\ghc-9.8.2\cabal-install-3.11.0.0\t\unit-tests\build\unit-tests\unit-tests.exe -j1 --hide-successes (87/104 sec, 1)
```

jasagredo commented:
^ This was fixed in https://github.com/haskell/cabal/pull/9915. Needs a rebase

jasagredo commented:
All tests pass in my machine if using:
- Rebased on top of #9915 
- No symlinks
- CLANG64 environment
- Installed pkgconf from mingw-w64-clang64-x86_64-pkgconf

jasagredo commented:
FTR: I'm using [this](https://stackoverflow.com/questions/5917249/git-symbolic-links-in-windows) to undo-redo the symlinks on Windows

andreabedini commented:
I did the rebase but I some tests failed locally. Perhaps the code has bit-rotted a bit.

jasagredo commented:
You will have to add a step somewhere to install pkg-config on Windows:

> Cannot find pkg-config program. Cabal will continue without solving for

andreabedini committed:
Add separate cache for getPkgConfigDb

Querying pkg-config for the version of every module can be a very
expensive operation on some systems. This change adds a separate,
per-project, cache for PkgConfigDB; reducing the cost from "every plan
change" to "every pkg-config-db change per project".

The cache key is composed by the pkg-config configured program and the
list of directories reported by pkg-config's pc_path variable.

Co-authored-by: andreabedini <69135+andreabedini@users.noreply.github.com>

fix: skip MonitorPkgConfig test on Windows (no pkg-config available)

fix(MonitorPkgConfig): provide fake pkg-config for Windows CI

fix: address review comments
